tracer/src/Datadog.Trace.Tools.Runner/Utils.cs

```diff
@@ -251,30 +253,38 @@ public static async Task<bool> CheckAgentConnectionAsync(string agentUrl)
 
             var tracerSettings = new TracerSettings(configurationSource);
             var settings = tracerSettings.Build();
-            var discoveryService = DiscoveryService.Create(settings.Exporter);
-
-            var agentWriter = new ApmAgentWriter(settings, new CISampler(), discoveryService);
-
-            try
-            {
-                if (!await agentWriter.Ping().ConfigureAwait(false))
+            var discoveryService = new DiscoveryService(
+                AgentTransportStrategy.Get(
+                    settings.Exporter,
+                    productName: "discovery",
+                    tcpTimeout: TimeSpan.FromSeconds(5),
+                    AgentHttpHeaderNames.MinimalHeaders,
+                    () => new MinimalAgentHeaderHelper(),
+                    uri => uri),
+                10,
+                1000,
+                int.MaxValue);
+
+            var tcs = new TaskCompletionSource<AgentConfiguration>(TaskCreationOptions.RunContinuationsAsynchronously);
+            discoveryService.SubscribeToChanges(
+                aCfg =>
                 {
-                    WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
-                    return false;
-                }
-            }
-            catch (Exception ex)
-            {
-                WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
-                AnsiConsole.WriteException(ex);
-                return false;
+                    tcs.TrySetResult(aCfg);
+                    discoveryService.DisposeAsync();
+                });
+
+            var cts = new CancellationTokenSource();
+            cts.CancelAfter(5000);
+            using (cts.Token.Register(
+                       () =>
+                       {
+                           WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
+                           tcs.TrySetResult(null);
+                           discoveryService.DisposeAsync();
+                       }))
+            {
+                return await tcs.Task.ConfigureAwait(false);
             }
```

Comment:
You might want to move the dispose _after_ the `await tcs.Task`, just so you can await it properly? I know doesn't _currently_ do anything actually `async`, but it might in the future, and this could end up a tricky thing to diagnose?

```suggestion
                    tcs.TrySetResult(aCfg);
                });

            var cts = new CancellationTokenSource();
            cts.CancelAfter(5000);
            using var t = cts.Token.Register(
                       () =>
                       {
                           WriteError($"Error connecting to the Datadog Agent at {tracerSettings.Exporter.AgentUri}.");
                           tcs.TrySetResult(null);
                       })

            var config = await tcs.Task.ConfigureAwait(false);
            await discoveryService.DisposeAsync();
            return config;

```